examples/bench/main.go

```diff
@@ -0,0 +1,1020 @@
+		return benchmarks(screen)
+	}
+	if err := ebiten.Run(f, 320, 240, 1, "Test"); err != nil && err != noErr {
+		panic(err)
+	}
+	for _, b := range benchList {
+		fmt.Printf("%s\n", b.name)
+		fmt.Println(b.result)
+		fmt.Println(b.result.MemString())
+	}
+}
+
+var (
+	px26  *ebiten.Image
+	px104 *ebiten.Image
+	px416 *ebiten.Image
+)
+
+var benchList = []benchmark{
+	{
+		name: "draw26",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px26, nil)
+			}
+		},
+	},
+	{
+		name: "draw104",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px104, nil)
+			}
+		},
+	},
+	{
+		name: "draw416",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px416, nil)
+			}
+		},
+	},
+	{
+		name: "draw26color1",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			ops := []*ebiten.DrawImageOptions{&ebiten.DrawImageOptions{}, &ebiten.DrawImageOptions{}}
+			ops[0].ColorM.Scale(1.0, 0.2, 0.2, 1.0)
+			ops[1].ColorM.Scale(0.2, 1.0, 0.2, 1.0)
+			for i := 0; i < b.N; i++ {
+				idx := i
+				ops[idx%2].GeoM.Reset()
+				ops[idx%2].GeoM.Translate(float64(i*320/b.N), float64(i*200/b.N))
+				screen.DrawImage(px26, ops[idx%2])
+			}
+		},
+	},
+	{
+		name: "draw26color10",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			ops := []*ebiten.DrawImageOptions{&ebiten.DrawImageOptions{}, &ebiten.DrawImageOptions{}}
+			ops[0].ColorM.Scale(1.0, 0.2, 0.2, 1.0)
+			ops[1].ColorM.Scale(0.2, 1.0, 0.2, 1.0)
+			for i := 0; i < b.N; i++ {
+				idx := i / 10
+				ops[idx%2].GeoM.Reset()
+				ops[idx%2].GeoM.Translate(float64(i*320/b.N), float64(i*200/b.N))
+				screen.DrawImage(px26, ops[idx%2])
+			}
+		},
+	},
+}
+
+var (
+	benchIdx  = 0
+	results   []testing.BenchmarkResult
+	didSetup  sync.Once
+	benchFunc func(b *testing.B, screen *ebiten.Image)
+	benchGo   = make(chan struct{})
+	benchDone = make(chan bool)
+)
+
+func benchmarks(screen *ebiten.Image) error {
+	var setupErr error
+
+	didSetup.Do(func() {
+		eimg, _, err := openEbitenImage()
+		if err != nil {
+			setupErr = fmt.Errorf("can't open image: %s", err)
+			return
+		}
+		px26 = eimg
+
+		w, h := eimg.Size()
+		img, err := ebiten.NewImage(w*4, h*4, ebiten.FilterNearest)
+		if err != nil {
+			setupErr = fmt.Errorf("can't create ne image: %s", err)
+			return
+		}
+		px104 = img
+
+		op := &ebiten.DrawImageOptions{}
+		op.GeoM.Scale(4, 4)
+		px104.DrawImage(px26, op)
+
+		img, err = ebiten.NewImage(w*16, h*16, ebiten.FilterNearest)
+		if err != nil {
+			setupErr = fmt.Errorf("can't create ne image: %s", err)
+			return
+		}
+		px416 = img
+		op.GeoM.Scale(4, 4)
+		px416.DrawImage(px26, op)
+	})
+	if setupErr != nil {
+		return setupErr
+	}
+
+	if benchFunc != nil {
+		benchGo <- struct{}{}
+		if <-benchDone {
+			benchFunc = nil
+		}
+	} else {
+		if len(benchList) > benchIdx {
+			benchFunc = benchList[benchIdx].fn
+			go func(idx int) {
```

Comment:
Not sure why this code is called asyncly.

Reply:
Because the Benchmark() call has to run over a period of 60 frames, and thus can't run synchronously inside a single frame's update.